File: ts/src/algorithm.ts
import Student from './student'
import Seminar from './seminar'

enum Points {
  FIRST_SELECTION = 0,
  SECOND_SELECTION = 5,
  THIRD_SELECTION = 10,
  NO_SELECTION = 30
}

export type Iteration = {
  id: number,
  points: number,
  assignments: Map<Student, { wSeminar?: Seminar, pSeminar?: Seminar }>
}

function shuffle<T>(a: Array<T>) {
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]]
  }

  return a
}

export default async function runAlgorithm(iterations: number, students: Student[]): Promise<Iteration> {
  // TODO Multithreading
  return execute(iterations, students)
}

function tryAssignment(seminar: Seminar, student: Student, currentIteration: Iteration, seminarType: 'W' | 'P', points: Points) {
  let cap = seminar.remainingCapacityPerRun.get(currentIteration.id)
  if (cap === undefined) {
    cap = seminar.capacity
  }

  if (cap > 0) {
    seminar.remainingCapacityPerRun.set(currentIteration.id, cap - 1)
    currentIteration.points += points

    const previousAssignment = currentIteration.assignments.get(student) || {}
    if (seminarType === 'W') {
      currentIteration.assignments.set(student, {...previousAssignment, wSeminar: seminar})
    }

    if (seminarType === 'P') {
      currentIteration.assignments.set(student, {...previousAssignment, pSeminar: seminar})
    }

    student.pointsPerRun.set(currentIteration.id, (student.pointsPerRun.get(currentIteration.id) || 0) + points)

    return true
  }

  return false
}

function execute(iterations: number, students: Student[]): Iteration {
  let best: Iteration | null = null
  for (let i = 0; i < iterations; i++) {
    const currentIteration: Iteration = {id: i, points: 0, assignments: new Map()}
    const studentsCopy = shuffle([...students])
    for (const student of studentsCopy) {
      if (
        !tryAssignment(student.wWishes[0], student, currentIteration, 'W', Points.FIRST_SELECTION) &&
        !tryAssignment(student.wWishes[1], student, currentIteration, 'W', Points.SECOND_SELECTION) &&
        !tryAssignment(student.wWishes[2], student, currentIteration, 'W', Points.THIRD_SELECTION)
      ) {
        student.pointsPerRun.set(i, Points.NO_SELECTION)
        currentIteration.points += Points.NO_SELECTION
      }
    }

    studentsCopy.sort((a, b) => {
      return (b.pointsPerRun.get(i) || 0) - (a.pointsPerRun.get(i) || 0)
    })

    for (const student of studentsCopy) {
      if (
        !tryAssignment(student.pWishes[0], student, currentIteration, 'P', Points.FIRST_SELECTION) &&
        !tryAssignment(student.pWishes[1], student, currentIteration, 'P', Points.SECOND_SELECTION) &&
        !tryAssignment(student.pWishes[2], student, currentIteration, 'P', Points.THIRD_SELECTION)
      ) {
        student.pointsPerRun.set(i, Points.NO_SELECTION + (student.pointsPerRun.get(i) || 0))
        currentIteration.points += Points.NO_SELECTION
      }
    }

    if (best === null || best.points > currentIteration.points) {
      best = currentIteration
    }
  }

  if (!best) {
    throw new Error('Minimum of 1 iteration required')
  }

  return best
}

File: python/Algorithm.py
import random
from random import shuffle
from typing import List

from Iteration import Iteration
from Seminar import Seminar, SeminarType
from Student import Student


def try_assignment(iteration: Iteration, student: Student, seminar: Seminar, seminar_type: SeminarType,
                   selection_points: int) -> bool:
    cap = iteration.seminarCapacities.get(seminar, seminar.capacity)

    if cap > 0:
        iteration.seminarCapacities[seminar] = cap - 1
        iteration.assignments[student][seminar_type] = seminar
        iteration.points_per_student[student] += selection_points
        return True

    return False


def run_algorithm(students: List[Student], iterations: int) -> Iteration:
    points = {
        'FIRST_SELECTION': 0,
        'SECOND_SELECTION': 5,
        'THIRD_SELECTION': 10,
        'NO_SELECTION': 30
    }

    best_iteration: Iteration | None = None

    for i in range(iterations):
        iteration = Iteration()
        iteration.points_per_student = {student: 0 for student in students}

        random.shuffle(students)

        for student in students:
            iteration.assignments[student] = {}

            if (
                    not try_assignment(iteration, student, student.w_wishes[0], SeminarType.SCIENTIFIC,
                                       points['FIRST_SELECTION']) and
                    not try_assignment(iteration, student, student.w_wishes[1], SeminarType.SCIENTIFIC,
                                       points['SECOND_SELECTION']) and
                    not try_assignment(iteration, student, student.w_wishes[2], SeminarType.SCIENTIFIC,
                                       points['THIRD_SELECTION'])
            ):
                iteration.points_per_student[student] += points['NO_SELECTION']

            if (
                    not try_assignment(iteration, student, student.p_wishes[0], SeminarType.PRACTICAL,
                                       points['FIRST_SELECTION']) and
                    not try_assignment(iteration, student, student.p_wishes[1], SeminarType.PRACTICAL,
                                       points['SECOND_SELECTION']) and
                    not try_assignment(iteration, student, student.p_wishes[2], SeminarType.PRACTICAL,
                                       points['THIRD_SELECTION'])
            ):
                iteration.points_per_student[student] += points['NO_SELECTION']

        if best_iteration is None or iteration.total_points() < best_iteration.total_points():
            best_iteration = iteration

    if best_iteration is None:
        print('No iteration found')
        exit(-1)
    return best_iteration
